### DISCUSSION TITLE
Squash merge?

### DISCUSSION THREAD

**Discussion by jrbourbeau** [earthaccess-dev/earthaccess#892] [General]
I don't feel super strongly, but have a mild preference for using squash merging on most projects. This makes it so contributors don't have to be careful with individual commits (e.g. `git commit -m "typo"`, `git commit -m "Ah, forgot to lint"`) in their PRs and the `git log` on `main` is relatively clean (one commit per PR). 

Thoughts from others on moving to sqash merging?

**Comment by MattF-NSIDC** [comment]
I prefer to use squash merging on a case-by-case basis. In PRs with well-curated commits, I like to keep them as-is. When we get PRs with poorly-curated commits, I'm all for a squash merge!

**Comment by jhkennedy** [comment]
I personally don't think squash merges solve anything but vanity^ -- but I also don't really *care* what's the norm here, so squashes are fine by me if that's the preference. 

I only have 2 actual concerns, and that's true with any merge method:
1. I want the green merge button by default to just do what the norm is so maintainers don't really have to think about it (helps expand the maintainer team!)
2. CI/CD often comes with assumptions around how things are merged, so we just need to make sure it's consistent with the allowed merge methods

so it might be better to just pick one of these, and dis-allow the rest:
![image](https://github.com/nsidc/earthaccess/assets/7882693/be8e0613-3508-433d-bebd-7277d22a38eb)

----

^ It can be argued that squash merges make git bisect easier, but you lose the journey/context that's often helpful for debugging without GitHub, so offline debugging is harder. Overall the positives and the negatives end up in the "meh" category for me and it ends up just being a discussion of what's prettier.

**Comment by MattF-NSIDC** [comment]
As I'm thinking about this I keep changing my mind. I don't think there's any "correct" decision here... 

I think "is a clean history important?" is the root question, and I think "yes", but not above all, especially not above community health or inclusivity. I like a clean history, and I like to curate my commits to the best of my ability (after a particularly bad day, that ability can leave my body :laughing:), but I don't think we should leverage that expectation on contributors.

I think the best solution may just to be loosey-goosey about it, make it a "best effort" kind of thing and it'll work out fine. Encourage merge commits for well-curated PRs and squashes for messier ones, but be OK with people selecting the default and sometimes having messy PRs not get squashed. You can still get a clean one-commit-per-PR log view of `main` with merge commits.

The strongest opinion I have on this, I think, is that we should disable rebase merging, because even with well-curated commits, I'd rather only see the merge commit on `main` than all of them. Rebasing is also not nearly as broadly understood as merging, and I remember being intimidated by rebasing myself at one time :)

**Comment by jhkennedy** [comment]
>  I don't think there's any "correct" decision here...

:100: 

> The strongest opinion I have on this, I think, is that we should disable rebase merging, because even with well-curated commits, I'd rather only see the merge commit on main than all of them. Rebasing is also not nearly as broadly understood as merging, and I remember being intimidated by rebasing myself at one time :)

I can agree to that and am happy to uncheck that one at least :smile:

**Comment by mfisher87** [comment]
@yuvipanda makes an excellent point here: https://github.com/jupyterhub/team-compass/issues/686#issuecomment-1767629271

And thinking back on how I merged #566, I apologize @botanical for not considering how the choice might affect you. Personally, Yuvi's argument has swayed me and now I have stronger feelings :)

**Comment by mfisher87** [comment]
I feel like a good outcome from this ticket may be either:

* Pick one strategy and disallow others (simple, not nuanced)
* Disable "rebase" strategy and write a documentation page/section on our merge-vs-squash policy and our motivations behind it -- e.g. "merge by default, sometimes squash - be wary of power dynamics and let the contributor's preference take precedence over the maintainer's"? (not simple, nuanced)

I feel like the latter may be the right way for us, and I know a lot of people are pro-squash... happy to defer if nobody else is swayed, but I do think inclusiveness is more important than cleanliness here.

**Comment by chuckwondo** [comment]
My OCD tendencies generally lead me to keeping my git histories "cleaner" (for better or worse), but I am happy to agree on the "merge by default, sometimes squash" approach here.

**Comment by JessicaS11** [comment]
Personally, I prefer squashing because it means I'm more likely to commit more often (e.g. at the end of the workday I can say "WIP blah blah" because my half-baked idea won't be part of the main/dev branch history). This means I worry less about "airing my dirty laundry" while I'm figuring things out. As a maintainer who's not a git wizard, I also find it far easier to see how the library has progressed if I can read the history without every single "fix typo" commit showing up. It also helps if others aren't writing good commit messages...

All that said, I support @MattF-NSIDC's second bullet:
> Disable "rebase" strategy and write a documentation page/section on our merge-vs-squash policy and our motivations behind it -- e.g. "merge by default, sometimes squash - be wary of power dynamics and let the contributor's preference take precedence over the maintainer's"? (not simple, nuanced)

Is there a way to have a required checkbox when a PR is created that allows the contributor to easily specify which type of merge they'd prefer? I'm thinking about how we don't end up with approved PRs hanging until the contributor comes back and merges.

**Comment by jhkennedy** [comment]
> Personally, I prefer squashing because it means I'm more likely to commit more often (e.g. at the end of the workday I can say "WIP blah blah" because my half-baked idea won't be part of the main/dev branch history). 

@JessicaS11 I would *not* hesitate to commit whether we're doing squash merges or merge commits; development is messy, history is messy, and it's fine. 

> As a maintainer who's not a git wizard, I also find it far easier to see how the library has progressed if I can read the history without every single "fix typo" commit showing up. It also helps if others aren't writing good commit messages...

That's a good point; there are a few commands and aliases you can use to filter for this with git to reduce the clutter you have to filter through. 

I don't have them on hand since I don't mind the mess when I look at git history (my team at ASF uses merge commits, and we don't bother rebasing/squashing out those things), but I'll circle back with them, which we could document.

**Comment by jhkennedy** [comment]
> Is there a way to have a required checkbox when a PR is created that allows the contributor to easily specify which type of merge they'd prefer?

We could add it to the template, or automate a comment on PRs asking, but I don't think there's a robust way to *require* it. I also think, generally, that puts too much on contributors as "Would you prefer a merge commit or a squash merge?" requires quite a bit of git knowledge to make an informed decision.

**Comment by mfisher87** [comment]
> > Is there a way to have a required checkbox when a PR is created that allows the contributor to easily specify which type of merge they'd prefer?
> 
> We could add it to the template, or automate a comment on PRs asking, but I don't think there's a robust way to _require_ it. I also think, generally, that puts too much on contributors as "Would you prefer a merge commit or a squash merge?" requires quite a bit of git knowledge to make an informed decision.

I really like these ideas. I think we can make it easy on contributors by treating it like a consent request in the PR template. By default, if they don't take any action, we don't squash. If we feel the need to do changes to the history in a PR, we can discuss it with the contributor as usual!

> - [ ] *OPTIONAL* Check this box by changing `[ ]` to `[x]` at the start of the line if you are OK with having your commits "squashed" when this change is accepted. This helps us keep a clean history by combining many commits in to one, but some contributors may prefer to keep their commits as-is (for example, one commit adds a failing test to reproduce a bug, another commit fixes the bug), and that's OK.

Thoughts?

**Comment by mfisher87** [comment]
I think since I last posted on this issue, I was pro-squash, but OK mixing methods. Now I'm more squarely pro-squash and I'd like to disable merge commits. We're pretty much only doing merge commits right now.

I think this is really inaccessible (this is earthaccess' actual history on my local):

<img width="1896" height="1620" alt="image" src="https://github.com/user-attachments/assets/f3dc44f3-35e2-454b-b109-23c31cdc9f61" />

And this option is very accessible:

<img width="2394" height="1720" alt="image" src="https://github.com/user-attachments/assets/21ea5e4f-faca-41d3-ba0c-c9775e99eebc" />


For an audience who might be new to Git, I'd much rather have them see the latter than the former. I don't agree with "the commit is the unit of work" argument -- perhaps this is true with very disciplined developers, and I try to work this way, but I don't always. I think in the way we work, the PR is the unit of work. Very rarely do I look at specific commits even when reviewing. If a developer wants to be disciplined and curate their commits, that's great! We can still see those in the PR interface if we want. The commits are more useful as a unit of work within a PR than they are as a unit of work in the larger scope of the project. As a consumer of the _project_ history, I don't care if you added failing unit tests first then implemented your feature. I might care in the scope of a PR, as a reviewer, and squashing preserves that experience.

>I prefer squashing because it means I'm more likely to commit more often (e.g. at the end of the workday I can say "WIP blah blah" because my half-baked idea won't be part of the main/dev branch history
> [...]
> As a maintainer who's not a git wizard, I also find it far easier to see how the library has progressed if I can read the history without every single "fix typo" commit showing up. It also helps if others aren't writing good commit messages...

We can only provide the kind of experience @JessicaS11 is looking for if we move to a 100% squash model. I don't think we should consider rewriting our history, we should just move forward with the new model.

**Reply by mfisher87** [reply]
cc @chuckwondo @jhkennedy @JessicaS11

**Reply by jhkennedy** [reply]
For features into `main`, I honestly don't care one way or the other, so whatever y'all decide is fine. 

I prefer merge commits and don't have any trouble navigating the history personally so most of the arguments for squashing just sound vain to me. 

The only real drawback to squashing is that it disconnects my local git history from the project history. I have to go to GitHub to look at what happened in a PR. But, most all of our artifacts are in GitHub, I hardly ever work where I can't get to GitHub, and we take a lot of contributions from forks, so... that's not really a concern here.

However, fully automating the release process with "release" or "tag" requests, as discussed in https://github.com/earthaccess-dev/earthaccess/discussions/857, is a lot more challenging for "reasons"^ if we disable merge commits globally (in the general settings) vs requiring linear history as a branch protection rule for `main`. That said, we have not made any movement towards automating the release process in ~1.5 years, so we could just globally disable merge commits and revisit it if we do pick it up down the line.

And, superseding all of :point_up: for me is that I strongly prefer to have whatever is default when you open a PR as an okay way to merge -- maintainers shouldn't have to think about the type of merge and just be fine pushing the big merge button. Only having 1 way to merge definitely enforces that and is how I set up my projects (I just only allow merge commits :rofl: ).

---

^GitHub's UI sucks for trunk based development and the psudo release branch is a work around for either of the two missing UX features that would eliminate this complexity entirely (fast-forward merges which GitLab has had forever and/or a way to propose a tag/release).

**Reply by mfisher87** [reply]
>arguments for squashing just sound vain to me

I do want to specify that, as a Git nerd, I prefer well-curated commits and am not bothered by a complex history, so my argument isn't coming from "it looks better" even a little bit :)
